[System] Fixes for path manipulation with tests.

78andyp opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Following on from @CrystalP's #27488 I remembered I had started this a while ago but pulled out the bluray:// parts.

I've gone through almost all of the path manipulation routines and tried to make sure there are tests that cover DOS paths, linux paths, smb:// paths, bluray:// paths, zip:// paths, rar:// paths and archive:// paths.

Look at the tests and work backwards is probably easiest.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

1) zip:// and rar:// paths were not being properly stored in the database.
2) Metadata and external subs/audio were not being properly located for zip://, archive:// and stack:// paths.
3) archive:// paths were being wrongly cleaned from the database.
4) DOS paths were being inconsistently handled.
5) During stack:// path construction not all separators between movie name and part number were removed.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

1) Adding tests.
2) Functional tests using this directory structure:
```
├───28 Weeks Later (2007)
│       28.Weeks.Later.2007.1080p.BluRay.x264.DTS-FGT.rar
│       28.Weeks.Later.2007.1080p.BluRay.x264.DTS-FGT.zul.srt
│
├───The Hobbit An Unexpected Journey (2012)
│       The Hobbit An Unexpected Journey (2012) Part 1.mp4
│       The Hobbit An Unexpected Journey (2012) Part 2.mp4
│       The Hobbit An Unexpected Journey (2012) Part 3.mp4
│       The Hobbit An Unexpected Journey (2012) Part 4.mp4
│       The Hobbit An Unexpected Journey (2012).zul.srt
│
├───The Lord of the Rings The Two Towers (2002)
│   │   The Lord of the Rings The Two Towers (2002).zul.srt
│   │
│   ├───The Lord of the Rings The Two Towers (2002) part 1
│   │       The Lord of the RIngs The Two Towers Part 1.mp4
│   │
│   ├───The Lord of the Rings The Two Towers (2002) part 2
│   │       The Lord of the RIngs The Two Towers Part 2.mp4
│   │
│   ├───The Lord of the Rings The Two Towers (2002) part 3
│   │       The Lord of the RIngs The Two Towers Part 3.mp4
│   │
│   └───The Lord of the Rings The Two Towers (2002) part 4
│           The Lord of the RIngs The Two Towers Part 4.mp4
│
├───Van Wilder (2002)
│       Van.Wilder.2002.PROPER.1080p.BluRay.x265-RARBG.zip
│       Van.Wilder.2002.PROPER.1080p.BluRay.x265-RARBG.zul.srt
│
└───Van Wilder Freshman Year (2009)
        Van.Wilder.Freshman.Year.2009.UNRATED.1080p.AMZN.WEBRip.DDP5.1.x264-pawel2006.tar.gz
        Van.Wilder.Freshman.Year.2009.UNRATED.1080p.AMZN.WEBRip.DDP5.1.x264-pawel2006.zul.srt
```

Library scanned then archive and rar add-ons are enabled and library scanned again.

_**Before**_

**Path table**
 
<img width="940" height="174" alt="image" src="https://github.com/user-attachments/assets/0c2f76de-2f3e-4a2b-8a5d-242e775b0f8f" />

**File table**
 
<img width="940" height="121" alt="image" src="https://github.com/user-attachments/assets/900f6c86-97ee-44d4-8a82-1f72a47e0dc7" />

**External subtitles**
 
<img width="940" height="380" alt="image" src="https://github.com/user-attachments/assets/29fa4d6d-c82f-4848-abb7-ae6c3e1714f4" />

Only found for .rar files

**All play**

**Refresh**

Zip:// - works
Archive:// - works
Rar:// - works
Stack:// - crashes

**Export to NFO**

rar:// correct
stack:// (files) correct
stack:// (folders) nfo has no filename, art placed in folder of movie name. Art within has no filename.
zip:// correct
archive:// fails

_**After**_

**Path table**
 
<img width="2403" height="533" alt="image" src="https://github.com/user-attachments/assets/7107194d-70b8-48cf-8ff6-d7b8bb4d82c6" />

Note lower case encoding in rar:// path,.

**File table**

<img width="940" height="123" alt="image" src="https://github.com/user-attachments/assets/f7c07979-bad0-4101-ac82-0927b5256468" />
 
**External subtitles**
 
<img width="940" height="459" alt="image" src="https://github.com/user-attachments/assets/0135d4e0-b2d5-4f17-b6dd-a5595bf9b373" />

Found for all

**All play**

**Refresh**

Zip:// - works
Archive:// - works
Rar:// - works
Stack:// - works

**Export to NFO and subsequent import (scan with local scraper)**

rar:// correct
stack:// (files) correct
stack:// (folders) correct
zip:// correct
archive:// correct

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

78andyp committed:
[Tests] Add tests.

78andyp committed:
[VideoDatabase] Fix path handling of rar:// and zip:// files.

The entire rar:// path was being returned as the filename, which seems against convention of the filename just being the media file itself.
The directory element within the archive was being lost (eg. rar://xxx/movie.rar/BDMV/index.bdmv was being reduced to rar://xxx/movie.rar/ (path) and index.bdmv (file))
Some files, whose paths were generated by a VFS addon, could have encoding with capitals (eg. %2A instead of %2a) - this can lead to file/path mismatches in the database.

78andyp committed:
[FileItem] Fix wrong movie name derivation in GetMovieName() and GetMovieBasePath() means unable to scrape or refresh movie from information dialog with rar://, zip:// or stack:// paths.

78andyp committed:
[URIUtils] Fix CURIUtils::GetBasePath() to properly handle archive (eg. zip://, rar://, archive://, and stack://) paths.

78andyp committed:
[VideoDatabase] Fix m_basepath not populated in CVideoDatabase::GetFileInfo().

Commit:
[URIUtils] Fix lack of archive support in GetParentPath().

Optimise function for discs.

Commit:
[Util] Fix lack of support for archives and stacks in GetVideoBasePathAndFileName().

Commit:
[FileItem] Add comment for GetLocalMetaDataPath().

Commit:
[URIUtils] Fixes for archive:// and associated extensions like .tar.gz.

Also deals with the fact that the URIUtils extension functions don't handle extensions with two dots like .tar.gz.
This consolidates all extension related routines in URIUtils (URL functions now use those) with one base function (FindExtension) in an anonymous namespace.

Commit:
[Util] Fix archive:// always cleaned from library.

Optimise for stack://.

78andyp commented:
@claude - looking at the commits in this PR and the build failures - focusing on:
```/home/jenkins/workspace/FreeBSD/xbmc/utils/test/TestURIUtils.cpp:420
Expected equality of these values:
  ref.c_str()
    Which is: "D:\\Movies\\"
  var.c_str()
    Which is: ""
```
Can you suggest why OSX and some linux builds may fail on DOS paths??

Commit:
[URIUtils][URL] Fix inconsistent handling of DOS/POSIX paths.

Commit:
[Util] Fix CUtil::GetPartNumberFromPath() fails on bluray:// paths.

Commit:
[StackDirectory] Fix DOS path handling in stack:// paths.

Alter function naming.

Improve folder regex to handle separators between movie name and part number.

Commit:
[VideoInfoScanner] Avoid unnecessary path processing for stacks in AddVideo().

78andyp commented:
@claude review

78andyp commented on xbmc/FileItem.cpp:
```diff
@@ -1898,22 +1881,26 @@ std::string CFileItem::GetBaseMoviePath(bool bUseFolderNames) const
+    if (name.empty())
+      name = strMovieName;
+    if (bUseFolderNames)
+      URIUtils::GetParentPath(name, strMovieName);
```
Done

Commit:
[Util] Fix external subtitles not found for stacks, zip:// and archive:// paths.

Commit:
[VideoDatabase] Fix Information -> Refresh crashes on stacks.

Commit:
[SaveFileStateJob] Fix duplicate file/path database entries can be created for archive files.

Commit:
[Util] Tidy Util.cpp includes.

78andyp commented:
@claude review latest additions.

Commit:
[ArtUtils] Fix ArtUtils for folder stacks and archives.

78andyp commented:
@claude review recent changes since last review.

Commit:
[URIUtils] Refactor SplitPath() and GetCommonPath()

78andyp commented:
jenkins build this please

Commit:
[NFOs] Fix nfos not found for folder stacks and some archives.

78andyp commented:
@claude review

garbear reviewed (approved):
Thanks for backing the fixes with so many tests. I haven't runtime tested but I didn't see any problems in the code.

CrystalP commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -493,12 +493,13 @@ void CAdvancedSettings::Initialize()
-  m_pictureExtensions =
-      ".png|.jpg|.jpeg|.bmp|.gif|.ico|.tif|.tiff|.tga|.pcx|.cbz|.zip|.rss|.webp|.jp2|.apng|.avif|.heif|.heic";
+  m_pictureExtensions = ".png|.jpg|.jpeg|.bmp|.gif|.ico|.tif|.tiff|.tga|.pcx|.cbz|.zip|.rss|.webp|."
+                        "jp2|.apng|.avif|.heif|.heic";
```
I didn't spot the change. Please revert to original if there isn't one.

CrystalP commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -493,12 +493,13 @@ void CAdvancedSettings::Initialize()
-                          ".ass|.vtt|.idx|.ifo|.zip|.sup";
+  m_subtitlesExtensions = ".utf|.utf8|.utf-8|.sub|.srt|.smi|.rt|.txt|.ssa|.text|.ssa|.aqt|.jss|."
+                          "ass|.vtt|.idx|.ifo|.zip|.sup";
+  m_archiveExtensions = ".tar.bz2|.tar.gz|.tar.xz|.tgz|.tbz2|.xz|.7z|.bz2|.gz|.tar|.rar|.zip";
```
do composed extensions like .tar.bz2 really work? The HasExtension() & cie functions do not handle them well.

CrystalP commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -148,128 +153,172 @@ std::string URIUtils::GetExtension(const CURL& url)
+  // Extract filename
+  const std::string filename{StringUtils::ToLower(path.substr(trim))};
+  if (filename.find_last_of('.') == std::string::npos)
+    return NO_EXTENSION; // No extension (a leading dot only)
```
I may have read too quickly, but variable period already contains the position of the last period. If it was determined to be at the start of the filename, instead of ignoring it we could have concluded NO_EXTENSION without looking for the last period again?

CrystalP commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -148,128 +153,172 @@ std::string URIUtils::GetExtension(const CURL& url)
+
+namespace
+{
+int FindExtension(const std::string& path,
```
The function is not trivial, a doxygen comment would be good.

CrystalP commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -148,128 +153,172 @@ std::string URIUtils::GetExtension(const CURL& url)
+        mask += "|.app|.applescript|.workflow";
 #endif
-    strFileMask += "|";
+        return std::string_view{mask};
```
static is actually wrong - the user can load overrides of the extensions list through advanced settings, and the extensions (with possibly different overrides) are reloaded when the user switches profile. And I don't think there is a guarantee that there can't be a call to RemoveExtension before adv settings are loaded.
The concatenated string could be cached, but the cache has to react to profile changes or other reasons for adv.settings to be reloaded. I have a PR that creates an event for such things. I don't know if this is called often and could justify the effort.

CrystalP commented on xbmc/utils/URIUtils.h:
```diff
@@ -223,6 +223,7 @@ class URIUtils
   static bool IsCDDA(const std::string& strFile);
   static bool IsDAV(const std::string& strFile);
   static bool IsDOSPath(const std::string &path);
+  static bool IsPOSIXPath(const std::string& path);
```
I understand you followed the pattern of IsDOSPath, but IsPosixAbsolutePath would be clearer, since relative paths return false but can still be posix paths.

CrystalP commented on xbmc/Util.cpp:
```diff
@@ -1084,7 +1084,8 @@ std::string CUtil::ValidatePath(std::string path, bool bFixDoubleSlashes /* = fa
   {
-    StringUtils::Replace(path, '\\', '/');
+    if (!URIUtils::IsDOSPath(path))
+      StringUtils::Replace(path, '\\', '/');
```
I don't think so, on non-Windows platforms the separator is supposed to be / unconditionally and the condition is redundant for the Windows case.
Why this change?

78andyp replied:
Removed.

CrystalP commented on xbmc/utils/test/TestURIUtils.cpp:
```diff
@@ -56,27 +61,112 @@ TEST_F(TestURIUtils, GetDirectory)
 {
-  EXPECT_STREQ(".avi",
-               URIUtils::GetExtension("/path/to/movie.avi").c_str());
+  EXPECT_STREQ(".avi", URIUtils::GetExtension("/path/to/movie.avi").c_str());
```
I can see the original used EXPECT_STREQ but EXPECT_EQ doesn't work with string literals?

78andyp replied:
Changed

CrystalP commented on xbmc/utils/test/TestArtUtils.cpp:
```diff
@@ -45,9 +44,8 @@ std::string unique_path(const std::string& input)
-  std::transform(input.begin(), input.end(), std::back_inserter(ret),
-                 [&randchar](const char c) { return (c == '%') ? randchar() : c; });
+  ret.resize(input.size());
+  std::ranges::transform(input, ret.begin(), [&](char c) { return c == '%' ? randchar() : c; });
```
const char as the original please.

78andyp commented on xbmc/utils/test/TestURIUtils.cpp:
```diff
@@ -56,27 +61,112 @@ TEST_F(TestURIUtils, GetDirectory)
+  EXPECT_STREQ("", URIUtils::GetExtension("movie").c_str());
+  EXPECT_STREQ(".ext", URIUtils::GetExtension(".movie.ext").c_str());
+  EXPECT_STREQ(".tar.gz", URIUtils::GetExtension("movie.tar.gz").c_str());
 }
```
Done

Commit:
[VideoDatabase] Fix a zip:// can be duplicated in database as an archive:// if the Archive add-on is installed after initial library scan and folder is modified.

CrystalP reviewed (commented):
FindExtension comments

CrystalP reviewed (commented):
IsDOS/PosixPath comments

Commit:
[Tests] Refactor some tests so that so that FileItem is only constructed after CServiceBroker is initialised, so that FindExtension() doesn't crash, as it uses CServiceBroker::GetFileExtensionProvider().GetCompoundArchiveExtensions().

78andyp commented:
1) Moved
2) Changed comment. Code is correct.
3) Wrong. IsDOSPath() checks for a letter followed by a colon at the start or two backslashes at the start. A url would start xxx://
4) A DOS path would be encoded in the URL so this is correct.
5) The URL structure is identical.

78andyp commented:
@claude review current changes and my comments above

78andyp commented:
@claude - please review latest changes in light of your last comments. please limit your comments to code that is directly altered by this PR (ie. comments 7 and 9 are out of scope of this PR)

78andyp commented:
jenkins build this please

78andyp commented:
@claude review code in context of my last comment using your numbering. Please give an example for point 4.

78andyp commented:
@CrystalP - is there anything else you want to add before I commit??
@garbear - I know you approved a little while ago (thanks) - I've made a few changes from @CrystalP's/AI feedback - do you want to look again??

garbear reviewed (approved):
I went through commit-by-commit and didn't see any problems. Given the scope of this PR (23 commits) I wouldn't be surprised if some small edge case escaped the numerous tests, but that's what nightly releases are for. Let's get this in and follow up if any unaddressed problems remain.

CrystalP commented:
Would you take a look at this forum issue https://forum.kodi.tv/showthread.php?tid=384424, playback from the library of a video in a rar archive . This PR likely has something to do with it.
The interesting line of the large log is
`2026-02-11 19:35:33.347 T:660     debug <general>: CGUIWindowVideoBase::PlayItem: File 'F:\rar3\tv\Friends\Season.4\Friends.S04E22.1080p.BluRay.x264.XBMC-DIRFIX-HOMEMADE\Friends.S04E23.1080p.BluRay.x264-TENEIGHTY\rar:\\F%3A%2Frar3%2Ftv%2FFriends%2FSeason.4%2FFriends.S04E22.1080p.BluRay.x264.XBMC-DIRFIX-HOMEMADE%2FFriends.S04E23.1080p.BluRay.x264-TENEIGHTY%2Fteneighty-friendss04e23.rar\teneighty-friendss04e23.mkv' for library item 'videodb://tvshows/titles/433/4/13882?season=4&tvshowid=433&xsp=%7b%22limit%22%3a15%2c%22order%22%3a%7b%22direction%22%3a%22descending%22%2c%22ignorefolders%22%3a0%2c%22method%22%3a%22lastplayed%22%7d%2c%22rules%22%3a%7b%22and%22%3a%5b%7b%22field%22%3a%22inprogress%22%2c%22operator%22%3a%22true%22%2c%22value%22%3a%5b%5d%7d%5d%7d%2c%22type%22%3a%22tvshows%22%7d' doesn't exist.
`
The \rar:\\ in the middle seems... odd.

CrystalP commented on xbmc/utils/test/TestURIUtils.cpp:
```diff
@@ -154,75 +406,412 @@ TEST_F(TestURIUtils, SplitPathLocal)
+{
+  const std::string& path{GetParam().path};
+  const std::string& base{URIUtils::GetBasePath(path)};
+  EXPECT_EQ(base, GetParam().parent);
```
I've been fighting this a bit to add a new test case and I think there is a typo and that this is meant to be `GetParam().base`. 44 tests were passing incorrectly (everywhere parent and base are supposed to be different)
With an impact on correctness of GetBasePath, downstream code and persisted values...

CrystalP replied:
The tests I meant to add were {.path = "D:\\", .parent = "", .base = "D:\\"}
and     {.path = "/", .parent = "", .base = "/"}

78andyp replied:
Having a look now...

78andyp replied:
So - it looks like all the failures are <path>/Disc 1 with the Disc 1 bit not being removed. I am somewhat reassured that none of the other tests fail (ie. GetLocalMetaDataPath(), GetMovieBasePath() etc..) as they tend to be the functional ones. Potentially a stack:// with a path that includes Disc n will fail to resolve to a common base but this is a corner case. Looking at a fix now.
Also the functional testing done for the path PR (as shown in the description) didn't show any obvious issues.
